### PR TITLE
Add custom mapper errors

### PR DESCRIPTION
**Pull request by keith**
This allows consumers to catch individual mapper failure cases.

### CODE REVIEW

File: Sources/Mapper.swift [hunk 1]
```diff
@@ -332,7 +339,7 @@ public struct Mapper {
      */
     @warn_unused_result
     public func from<T>(field: String, transformation: AnyObject? throws -> T) rethrows -> T {
-        return try transformation(self.JSONFromField(field))
+        return try transformation(try? self.JSONFromField(field))
```
Comment (Reflejo) [inline Sources/Mapper.swift]: This would prevent `JSONFromField` from propagating `MissingFieldError`; is this ok?

File: Sources/Errors.swift [hunk 2]
```diff
@@ -4,7 +4,16 @@
 
  Custom implementations of Mappable, Convertible, or transformation functions can throw any error they desire
  */
-public struct MapperError: ErrorType {
+public enum MapperError: ErrorType {
+    case ConvertibleError(value: AnyObject?, type: Any.Type)
+    case CustomError(message: String)
```
Comment (Reflejo) [inline Sources/Errors.swift]: Maybe we should include the field here as an optional? Seems like other than the legacy case every error would have a field

File: Sources/Errors.swift [hunk 3]
```diff
@@ -4,7 +4,16 @@
 
  Custom implementations of Mappable, Convertible, or transformation functions can throw any error they desire
```
Comment (Reflejo) [inline Sources/Errors.swift]: Should we document the cases?

**Comment by Reflejo** [comment]
Fixes: https://github.com/lyft/mapper/issues/31

Looks good to me

File: Sources/Mapper.swift (hunk 1 above)
Comment (keith) [inline Sources/Mapper.swift]: I'd rather not swallow this error, but there are 2 problems here. First if we didn't, we'd be making a breaking API change from `AnyObject?` to `AnyObject`. Overall I'm ok with doing that at some point, but not just yet. Second we couldn't utilize `rethrows` here, which is also too bad. But the first one is the bigger problem for now.

**Comment by keith** [comment]
Just went back and updated all other function documentation about throws to mention the specific errors that can be thrown and the conditions under which they are thrown.

File: Sources/Mapper.swift [hunk 4]
```diff
@@ -80,19 +82,25 @@ public struct Mapper {
      - parameter key: The key to retrieve from the source data, can be an empty string to return the entire
```
Comment (sberrevoets) [inline Sources/Mapper.swift]: Note: parameter name is outdated. Would be nice to change all instances of `key` to `field` here.

**Comment by sberrevoets** [comment]
Looks good 👍

File: Sources/Mapper.swift (hunk 4 above)
Comment (keith) [inline Sources/Mapper.swift]: Will update that in an upcoming PR

**Pull request merged**